### PR TITLE
2.x: WIP removes anonymous inner classes.

### PR DESCRIPTION
**Pull request by SleimanJneidi**
This is a WIP. Since the changes are big, an incremental review and PR was recommended by @akarnokd 

 - Removes them  from flowable/observable/operators.
- Issue [#5150](https://github.com/ReactiveX/RxJava/issues/5150)

### CODE REVIEW

File: src/main/java/io/reactivex/internal/operators/flowable/BlockingFlowableMostRecent.java [hunk 1]
```diff
@@ -78,44 +78,46 @@ public void onNext(T args) {
...
-                            buf = value;
-                        }
-                        if (NotificationLite.isComplete(buf)) {
-                            throw new NoSuchElementException();
-                        }
-                        if (NotificationLite.isError(buf)) {
-                            throw ExceptionHelper.wrapOrThrow(NotificationLite.getError(buf));
-                        }
-                        return NotificationLite.getValue(buf);
+        private class Iterator implements java.util.Iterator<T> {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/BlockingFlowableMostRecent.java]: `final`

Comment (akarnokd) [reply src/main/java/io/reactivex/internal/operators/flowable/BlockingFlowableMostRecent.java]: and no private

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableBufferTimed.java [hunk 2]
```diff
@@ -367,23 +358,31 @@ public void run() {
...
+            w.schedule(new RemoveFromBuffer(b), timespan, unit);
         }
 
         @Override
         public boolean accept(Subscriber<? super U> a, U v) {
             a.onNext(v);
             return true;
         }
+
+        private final class RemoveFromBuffer implements Runnable {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableBufferTimed.java]: No private

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableCombineLatest.java [hunk 3]
```diff
@@ -557,4 +552,11 @@ public void requestOne() {
 
         }
     }
+
+    private final class SingletonArrayFunc implements Function<T, R> {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableCombineLatest.java]: No private

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableDelay.java [hunk 4]
```diff
@@ -125,5 +102,45 @@ public void cancel() {
             w.dispose();
         }
 
+        private final class OnNext implements Runnable {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableDelay.java]: No private

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableDelay.java [hunk 5]
```diff
@@ -125,5 +102,45 @@ public void cancel() {
...
+                this.t = t;
+            }
+
+            @Override
+            public void run() {
+                actual.onNext(t);
+            }
+        }
+
+        private class OnError implements Runnable {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableDelay.java]: No private but final

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableDelay.java [hunk 6]
```diff
@@ -125,5 +102,45 @@ public void cancel() {
...
+            public void run() {
+                try {
+                    actual.onError(t);
+                } finally {
+                    w.dispose();
+                }
+            }
+        }
+
+        private class OnComplete implements Runnable {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableDelay.java]: no private but final

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableDelaySubscriptionOther.java [hunk 7]
```diff
@@ -38,71 +38,90 @@ public void subscribeActual(final Subscriber<? super T> child) {
...
         child.onSubscribe(serial);
 
-        FlowableSubscriber<U> otherSubscriber = new FlowableSubscriber<U>() {
-            boolean done;
+        FlowableSubscriber<U> otherSubscriber = new DelaySubscriber(serial, child);
+
+        other.subscribe(otherSubscriber);
+    }
+
+    private final class DelaySubscriber implements FlowableSubscriber<U> {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableDelaySubscriptionOther.java]: no private

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableDelaySubscriptionOther.java [hunk 8]
```diff
@@ -38,71 +38,90 @@ public void subscribeActual(final Subscriber<? super T> child) {
...
 
             @Override
-            public void onNext(U t) {
-                onComplete();
+            public void cancel() {
+                s.cancel();
             }
+        }
 
+        private final class OnCompleteSubscriber implements FlowableSubscriber<T> {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableDelaySubscriptionOther.java]: No private

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableRefCount.java [hunk 9]
```diff
@@ -183,23 +170,54 @@ void doSubscribe(final Subscriber<? super T> subscriber, final CompositeDisposab
...
-                        if (subscriptionCount.decrementAndGet() == 0) {
-                            baseDisposable.dispose();
-                            // need a new baseDisposable because once
-                            // disposed stays that way
-                            baseDisposable = new CompositeDisposable();
-                        }
+        return Disposables.fromRunnable(new DisposeTask(current));
+    }
+
+    private final class DisposeConsumer implements Consumer<Disposable> {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableRefCount.java]: no private

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableReplay.java [hunk 10]
```diff
@@ -1200,4 +1088,178 @@ Node getHead() {
...
+            }
+
+            final SubscriberResourceWrapper<R> srw = new SubscriberResourceWrapper<R>(child);
+
+            observable.subscribe(srw);
+
+            co.connect(new DisposableConsumer(srw));
+        }
+
+        private final class DisposableConsumer implements Consumer<Disposable> {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableReplay.java]: no private

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableTimeoutTimed.java [hunk 11]
```diff
@@ -178,6 +157,27 @@ public void dispose() {
         public boolean isDisposed() {
             return worker.isDisposed();
         }
+
+        private final class TimeoutTask implements Runnable {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableTimeoutTimed.java]: no private

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableTimeoutTimed.java [hunk 12]
```diff
@@ -289,5 +279,35 @@ public void request(long n) {
         public void cancel() {
             dispose();
         }
+
+        private final class TimeoutTask implements Runnable {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableTimeoutTimed.java]: No private

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableUnsubscribeOn.java [hunk 13]
```diff
@@ -86,12 +86,14 @@ public void request(long n) {
...
-                    @Override
-                    public void run() {
-                        s.cancel();
-                    }
-                });
+                scheduler.scheduleDirect(new Cancellation());
+            }
+        }
+
+        private final class Cancellation implements Runnable {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableUnsubscribeOn.java]: No private

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableWindowTimed.java [hunk 14]
```diff
@@ -839,6 +829,19 @@ public void run() {
                 this.open = open;
             }
         }
+
+        private final class Completion implements Runnable {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableWindowTimed.java]: no private

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableWithLatestFrom.java [hunk 15]
```diff
@@ -138,4 +116,34 @@ public void otherError(Throwable e) {
             actual.onError(e);
         }
     }
+
+    private final class FlowableWithLatestSubscriber<U> implements FlowableSubscriber<U> {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableWithLatestFrom.java]: no private

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableWithLatestFromMany.java [hunk 16]
```diff
@@ -298,4 +293,11 @@ public void dispose() {
             SubscriptionHelper.cancel(this);
         }
     }
+
+    private final class SingletonArrayFunc implements Function<T, R> {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableWithLatestFromMany.java]: no private

**Review by akarnokd — COMMENTED** [review]
Looks good so far, minus a couple of `private` modifiers that need to be removed.

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableDelaySubscriptionOther.java (hunk 7 above)
Comment (akarnokd) [reply src/main/java/io/reactivex/internal/operators/flowable/FlowableDelaySubscriptionOther.java]: Still private

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableDelaySubscriptionOther.java [hunk 17]
```diff
@@ -38,71 +38,90 @@ public void subscribeActual(final Subscriber<? super T> child) {
...
+        public void onComplete() {
+            if (done) {
+                return;
+            }
+            done = true;
+
+            main.subscribe(new OnCompleteSubscriber());
+        }
+
+        private class DelaySubscription implements Subscription {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableDelaySubscriptionOther.java]: no private but final

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableReplay.java (hunk 10 above)
Comment (akarnokd) [reply src/main/java/io/reactivex/internal/operators/flowable/FlowableReplay.java]: Here too.

**Review by akarnokd — COMMENTED** [review]
A couple of places still require attention.

**Comment by SleimanJneidi** [comment]
@akarnokd on it

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableDelay.java (hunk 5 above)
Comment (akarnokd) [reply src/main/java/io/reactivex/internal/operators/flowable/FlowableDelay.java]: Here too.

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableDelay.java (hunk 6 above)
Comment (akarnokd) [reply src/main/java/io/reactivex/internal/operators/flowable/FlowableDelay.java]: Here too.

**Review by akarnokd — COMMENTED** [review]
Still two missing. Look at the main page where there should be a bunch of collapsed comment along with the remaining problems in `FlowableDelay`.

**Comment by akarnokd** [comment]
Great job! I'll merge this so the changes don't get too numerous.

**Pull request merged**